setup-homebrew/main.sh

```diff
@@ -44,68 +44,96 @@ HOMEBREW_PREFIX="$(brew --prefix)"
 HOMEBREW_REPOSITORY="$(brew --repo)"
 HOMEBREW_CORE_REPOSITORY="$HOMEBREW_REPOSITORY/Library/Taps/homebrew/homebrew-core"
 HOMEBREW_CASK_REPOSITORY="$HOMEBREW_REPOSITORY/Library/Taps/homebrew/homebrew-cask"
+HOMEBREW_OTHER_CASK_REPOSITORIES=(
+    "${HOMEBREW_REPOSITORY}/Library/Taps/homebrew/homebrew-cask-drivers"
+    "${HOMEBREW_REPOSITORY}/Library/Taps/homebrew/homebrew-cask-fonts"
+    "${HOMEBREW_REPOSITORY}/Library/Taps/homebrew/homebrew-cask-versions"
+)
+if [[ "$GITHUB_REPOSITORY" =~ ^.+/homebrew-.+$ ]]; then
+    HOMEBREW_TAP_REPOSITORY="$(brew --repo "$GITHUB_REPOSITORY")"
+fi
 
 # Do in container or on the runner
 if [[ -f "/.dockerenv" ]] || ([[ -f /proc/1/cgroup ]] && grep -qE "actions_job|docker" /proc/1/cgroup); then
     # Fix permissions to give normal user access
-    sudo chown -R "$(whoami)" "$HOME" "$PWD/.."
+    if [[ -n "$(command -v setfacl)" ]]; then
+        setfacl_dirs=()
+        user=$(whoami)
+        for dir in "$HOME" "$RUNNER_WORKSPACE" "$RUNNER_TEMP"
+        do
+            if [[ ! -w "$dir" ]]; then
+                sudo setfacl -Rm "d:u:$user:rwX,u:$user:rwX" "$dir"
+                setfacl_dirs+=("$dir")
+            fi
+        done
+        echo "SETFACL_DIRECTORIES=$(IFS=:; echo "${setfacl_dirs[*]}")" >> $GITHUB_STATE
+        echo "Set up ACL."
+    elif [[ ! -w "$GITHUB_OUTPUT" ]]; then
+        echo "::warning::Missing write permissions to GitHub directories. Install \`acl\` (\`setfacl\`)."
+    fi
+
+    # Add safe directories.
```

Comment:
What are safe directories? Again: can this get more comments explaining.

Reply:
Because of https://github.blog/2022-04-12-git-security-vulnerability-announced/, running as root and performing git actions for a directory owned by `linuxbrew` results in an error without this.

See https://github.com/actions/checkout/commit/dcd71f646680f2efd8db4afa5ad64fdcba30e748 for the equivalent change GitHub did for `actions/checkout`.

Reply:
Sorry, still don't understand what behaviour this changes.

Reply:
Thanks! When is this called as root?

Reply:
For our own Docker images: we mostly avoid it ourselves except https://github.com/Homebrew/actions/blob/master/.github/workflows/setup-commit-signing.yml#L20.

For non-Homebrew Docker images: most of them run as root. The `--user=linuxbrew` we have is unusual for the most part (but makes sense for Homebrew/core etc.)

Reply:
Ok, thanks.

> GitHub only actually support root in Docker: https://docs.github.com/en/actions/creating-actions/dockerfile-support-for-github-actions#user, so it makes sense to make this work as root.

Note at least internal to GitHub: the opposite is the case 😄.

Given no-one has complained about this I'm fine with including it but don't feel like it's super high priority?

Reply:
> Note at least internal to GitHub: the opposite is the case 😄.

`actions/checkout@v3.1.0` no longer works as non-root anymore. It can't be used with our Homebrew Docker images anymore, hence the `setup-homebrew.yml` CI change.